### ISSUE TITLE
COSMIC shortcut system

### ISSUE THREAD

**Issue by maria-komarova** [pop-os/cosmic-epoch#46]
Shortcut system for navigating around the desktop and for window management. 
![Shortcuts Final Proposal table](https://user-images.githubusercontent.com/57502897/191095328-d6f0d2f8-54ee-4732-9852-62440c3d19e8.png)

**Comment by soc** [comment]
I'm super-happy seeing that all window operations keep using the `Window` key.

I absolutely hated it that in old DEs half of the ops were using `Alt`.

**Comment by WillyWonksters** [comment]
Has shift-super-alt-arrows been considered?  I find that combination to be much more ergonomic than the shift-super-ctrl-arrows that is currently proposed for moving a window to another workspace.

**Comment by GrubBox** [comment]
> Bravo Team
---
_I think these new *keymaps* will allow intuitive movement of windows. This is the best feature of *CosmicOS* and perhaps the main reason for its popularity. *KDE* has a similar feature but only via a plug in and its very complex to learn. This my sweet spot between a full blown Window manager and pre configured DE._

**Comment by krmbzds** [comment]
I personally prefer adjustment mode. I don't care for tiling, love vim and hate holding down modifier keys. Removing adjustment mode will break habits of existing users like me.

In `Current Pop Shortcut` column moving & resizing with **HJKL** was omitted so I assume those features will be removed as well. In that case, I think that will be a step backwards in terms of ergonomy since keeping hands in the home-row is crucial for keyboard centric workflows.

| direct hotkey |  adjustment mode |
|---|---|
| single step (faster) |  multi-step (slower) |
| immediate feedback, no preview |  ability to preview before taking action |
| need to hold down a modifier while moving windows | one hand operation after entering adjustment mode,  no need to hold down modifiers to move windows |
| unnatural to switch between moving and resizing | easy to switch between moving and resizing during adjustment since it requires a single modifier key |
| no HJKL for move & resize, must use arrow keys | move & resize with HJKL, keep hands in home-row | 

Please reconsider this.

**Comment by maria-komarova** [comment]
Thanks for the feedback, that's good for us to know. We're still trying to find what's best for most of our users and what better serves multiple needs so it's helpful to get this feedback. 

One thing to note is that HJKL are staying for all the shortcuts. The omission is merely an accident, or rather something we take for granted by now.

**Comment by GrubBox** [comment]
When is Cosmic Launching ? We have been waiting for 2 years now and most of the application are outdated. Neovim is running on 0.6  while even Debian Stable has it on 0.7

**Comment by pietryszak** [comment]
How to change workspace number for every monitor separetly?

**Comment by maria-komarova** [comment]
So far we were only planning to have separate numbering when displays have independent workspaces so the numbering will be automatic in this case. If we find out there is more need to have numbering system that covers all workspaces, we'll need to figure out how best to include this functionality. From what I saw in other systems, things can quickly get disorienting.

**Comment by pietryszak** [comment]
@maria-komarova So in my own i3 config file I have something like that:

- Main workspace monitor I use meta/windows + 1,2,3..

- Second monitor meta/windows + ctrl + 1,2,3...

Easy to remember

**Comment by oKcerG** [comment]
This seems an improvement but I still have some questions/remarks:

My usual usage 2 screens: a laptop screen with one tabbed stack and ultrawide monitor with 2/3 tabbed stacks arranged left to right. I only use a single workspace but make heavy use of the tabs (they act as an always visible overview of my windows, workspaces hide that).

When I have a window in a stack with multiple tabs I want to move it to a stack adjacent to it, would it be possible here? 
When in a stack the current move window action create a new stack between the new one and the adjacent one (after painstakingly changing the tab order of the window if I there is a lot of tabs). That's not what I usually want.
I also don't want new stacks or resizing my other existing windows when moving a window.
Would you consider a shortcut (Super-Shift-Alt + arrows ?) to move the window as a new tab in the adjacent stack?
I currently do it with the mouse but it's finicky, a keyboard shortcut to do that would be way more productive.

This allows shortcuts to be logically grouped: Ctrl as a workspace modifier, Alt as a stack modifier, Ctrl-Alt as monitor modifier, resulting in the following clear intuitive summary:

- Super + arrows is focusing an adjacent window 
  - Super-Ctrl + arrows is focusing an adjacent workspace
  - Super-Alt + arrows is focusing an adjacent stack
  - Super-Ctrl-Alt + arrows is focusing an adjacent monitor ( :exclamation: in your proposed change it is moving a window to an adjacent monitor, but I believe that having Super-Shift for all moving actions makes more sense and is more intuitive)
- Super-Shift + arrows is moving a window
  - Super-Shift-Ctrl + arrows is moving a window to an adjacent workspace
  - Super-Shift-Alt + arrows is moving a window to an adjacent stack
  - Super-Shift-Ctrl-Alt + arrows is moving a window to an adjacent monitor
  
 In this usage, Super-Shift-Alt should also tread single windows as stacks: 
 If my windows are arranged like this  in 3 horizontal splits :
 [ A | B | C ] - D - [ E - F ]
 where [ A | B | C ] and [ E - F ] are stacks with respectively 3 and 2 tabbed windows and D is a single unstacked window.
 When in A and doing Super-Shift-Alt + left it should result in:
 [ B | C ] - [ D | A ] - [ E - F ]
 
For the edge cases ( :upside_down_face: ) : alt moving when at the boundary of a workspace should move to the next workspace if there is one available, if not to the next monitor if there is one available. In both cases it should move to an existing stack if present or move it as a full size window if no existing window is in the destination. If no adjacent workspace or window is there in the move direction, splitting the current stack like regular moving seems a sensible choice.

**Comment by isaac-8601** [comment]
(Note: I'm a former contributor so I can't guarantee that my info is 100% accurate with the current thinking-- so nothing I say is from any position of authority!)

The Super-Shift-arrow shortcuts would allow windows to move into or out of stacks, with no additional modifier needed. If you had a workspace with windows **[A|B|C] - [D|E]**, using Super-Shift-right on Window C would create **[A|B] - C - [D|E]** (with C not in a stack) and if hit again would create **[A|B] - [C|D|E]**.

There was no specific stack-related shortcut planned that would, in the example above, move windows between stacks without the middle step of having a window between the two stacks. In a way, Super-Shift can act to move windows between stacks, workspaces, and monitors, depending on where the focused window starts.

I could see cases where a specific shortcut move between stacks would be faster. for actions where you want to move a window from one stack to the far end of another stack. But, with the proposal of the Super-Shift-arrows action working as outlined, does that meet the need sufficiently? (Or, does that feel sufficient compared to current Pop!_OS shortcuts?) How does that feel?

---

Additional unsolicited thoughts, to examine the idea more fully:

If a new shortcut were to be added, its impact on groups might also be considered. Stacks are _kind of_ like groups, so I imagine a new shortcut could move windows between logical groups, too. (I'm imagining columns of windows where moved windows jump to adjacent columns whereas in the original proposal they would move between columns additionally.) Personally, to me it doesn't seem like something I'd want to have another shortcut for, though. I prefer having to memorize less shortcuts and accept having to press the arrows a couple more times if the overall system is flexible enough. Curious how others feel.

**Comment by maria-komarova** [comment]
Thanks for the feedback and for the suggestion @oKcerG. We've talked about the shortcuts and it seems possible to add Super+Shift+Alt+arrows to move directly to the stack. But I wanted to clarify a few things first. 

As mentioned above, it is possible to move a window to the stack with Super+Shift+arrows. If you have two stacks next to each other on a laptop display, for example and try to move a window from the stack on the left to the right one, it would take hitting the shortcut with the right arrow four times to move the right-most window from the left stack to the first positioned window in the right stack. So if we add the Super+Shift+Alt+arrows it would speed up the process but result in more shortcuts. What system are you currently using? Pop!_OS 22.04? Or have you tried COSMIC shortcuts in action?

You mentioned that you don't want new stacks or resizing your other existing windows when moving a window. There would be no new stacks, but could you clarify about resizing. Are there any particular situations when resizing becomes cumbersome?

**Comment by oKcerG** [comment]
Thank you for the answers and insightful questions!

> What system are you currently using? Pop!_OS 22.04? Or have you tried COSMIC shortcuts in action?

Pop!_OS 22.04, I didn't take the time to look into how to try the new COSMIC stuff, I'll try to do that with the alpha.

> The Super-Shift-arrow shortcuts would allow windows to move into or out of stacks, with no additional modifier needed. If you had a workspace with windows [A|B|C] - [D|E], using Super-Shift-right on Window C would create [A|B] - C - [D|E] (with C not in a stack) and if hit again would create [A|B] - [C|D|E].
[...]  In a way, Super-Shift can act to move windows between stacks, workspaces, and monitors, depending on where the focused window starts.

Yes, I understood that, and it acts like that with the current adjustment mode in PopOS.  I really don't want the intermediate state where there is three stacks : [A|B] - C - [D|E], one in the middle that I don't care for and the sizing down and then back up of the source stack is disturbing. I also don't like the fact that if I want to move A to the stack on the right I have to press right 4 times, 2 times where it ends up in states that I consider equivalent to the original one in my workflow ([ A | B | C ], [ B | A | C ], [ B | C | A]), one that I don't want (single A), and a final one that I do want.

I sometimes have 2/3 stacks of 5 windows.

> But, with the proposal of the Super-Shift-arrows action working as outlined, does that meet the need sufficiently? (Or, does that feel sufficient compared to current Pop!_OS shortcuts?) How does that feel?

That doesn't feel sufficient as I explained above. It would feel cumbersome and I don't think I'll use the shortcuts altogether, I'll end up sticking with my mouse to move windows but that's a shame. That's what I do currently but I don't like it.

> If a new shortcut were to be added, its impact on groups might also be considered. 

Maybe yes, one could argue that someone using groups wouldn't mind the flexibility of Super-Shift and changing between layouts. There might be some ambiguous situation with a "move to adjacent group" shortcut to. What should a left move do in this situation?
![image](https://github.com/pop-os/cosmic-epoch/assets/349268/8f28ee26-dcc4-4bc1-837f-39ea2fffcc07)

Move it to a row with the big window? to a column? I guess the most logical would be a row since the active window is originally in a row. But that doesn't feel like the only choice someone would want.

With my proposed move to stack shortcut the choice seems clear, it would move the active window to a stack with the big window.

> I prefer having to memorize less shortcuts and accept having to press the arrows a couple more times if the overall system is flexible enough. Curious how others feel.

I agree with you there.

> if we add the Super+Shift+Alt+arrows it would speed up the process but result in more shortcuts.

Indeed it will, but I see it more as enabling another workflow for people like me:
I won't use nested groups, I just want to have a fixed layouts with 2 or 3 stacks side by side and be able to move windows between them easily.
If my suggestion is implemented, I will rebind Super-Shift-Alt to Super-Shift and vice versa in the rare case that I do want to use nested groups.

> You mentioned that you don't want new stacks or resizing your other existing windows when moving a window. There would be no new stacks, but could you clarify about resizing. Are there any particular situations when resizing becomes cumbersome?

What I meant by new stacks was new column (which in my mind is a stack with a single window) : the C in the [A|B] - C - [D|E] transitory state mentioned by isaac-8601.
As for the resizing this is the [A|B] window column shrinking to make place for the new column with C and growing back to their original size when the C is moved to the right column. This feels jarring.

To summarize my use-case:
I want to be able to use COSMIC with a layout that stays mostly fixed once I set it up.
With my laptop on the move I don't use tiling, on my wide screen at work I have 2 columns with stacks, on my ultra-wide screen at home I have 3 columns with stacks. I want to be able to move a window from one column to the other without creating new columns.

**Comment by alapins** [comment]
Trying the pre-alpha in a VM and very much like the tiling!  You've created something really great, that resonates with me in a way that no videos on i3 or other tiling managers has.  Had the following thoughts, in 3 categories:

### Bugs 
Pre-Alpha, so not sure whether these are known, but thought I'd mention them in case you hadn't run across them.
1. Using the mouse vs the keyboard has different visual overlays, specifically group outlines are visible in keyboard but not mouse.  I think seeing the groups are very helpful.
2. Using Super+U to select a group of two windows then Super+Shift+2. Expected it to move both windows in the group, but it only moved the window I originally had focused. Not sure whether this should work or whether I shouldn't be able to use group selection outside of Move/Shift mode. 
3. Have to hold down Super+X while using the arrow keys to have window swap work, which is unexpected. 
 
### Mnemonics/Keybindings/UX
1. Sheet above says Swap windows is Super+X.  Might want to describe it as eXchanging windows (which is where I'm assuming the shortcut actually came from to keep the mnemonics clear).
2. I'd suggest Super+Ctrl for resizing the currently selected window/group (super = windows key on most keyboards, looking to Control the size of the currently selected window, as opposed to Shifting it to a different location).  I'd very much like to be able to specify the size of a window/group either in pixels directly, or as a fraction of the overall screen width/height (ex. /3 would make it 1/3 of the display width).
3. Super+Alt or Super+Shift+Alt could then be for Alternate shift/move, i.e. between virtual desktops.  If there were some keyboard  shortcuts to move between groups directly, the "move to a different monitor" would be made easier; if not then they'd still be accessible using Super+Shift+arrows. I recognize this changes the current custom of using Ctrl to move workspaces, but I think it would be easier to remember.
4. It would be nice to have some shortcuts in addition to the arrows to move things to a different location.  either letters or numbers, both for groups and windows/stacks.  Choosing a window/stack would move the window into that stack or create a new stack with the new and old window.  To make a stack using the keyboard currently you have to go to the destination window, Super+S to make it a stack, navigate back to the original window and then move it to the new stack.  Seems like stacks should be handled in the same way groups are, where each window is implicitly part of a stack, but the tab chrome for the stack doesn't show for stacks containing one element. This would also address the comment above about being able to move a window between columns without creating a new column as an intermediate step.

### New Features
1. It would be nice to be able to see minimized windows (just titles?) in the layout view, and to be able to eXchange a visible window with a non-visible window. This would allow something like stacks without the tab chrome.
2. I'd like, when hitting Super+Shift, to see the layout overlay immediately, rather than when I hit an arrow key.  Some layouts aren't immediately intuitive where the group boundaries are, and I'm not sure what the best move is to do what I want.  Being able to see the landscape before moving would be helpful.  This might also address the desire to have a modal layout view (perhaps with a toggle in settings to prevent new users from getting confused by it) so that hitting Super+Shift on their own would  move it to the layout mode.
3. I'd love to be able to set up a file with a layout and the programs that would go in each portion of a group/stack, a la tmux.  Ideally I'd be able to save the current layout with the current programs in each slot.

**Comment by pietryszak** [comment]
> @maria-komarova So in my own i3 config file I have something like that:
> 
> - Main workspace monitor I use meta/windows + 1,2,3..
> 
> - Second monitor meta/windows + ctrl + 1,2,3...
> 
> Easy to remember

@maria-komarova what is status of that. That shortcuts for changing workspaces for only second monitor will be added before alpha iso release?

**Comment by avatar4d** [comment]
The overall look at feel of Cosmic Epoch is fantastic. It's truly amazing how far you've come in such a short time. 

That said, I'm finding the new default workflow for moving windows with the keyboard to be inefficient. In Gnome and the current Cosmic desktop, I can snap a window to the left or right side of my monitor and then quickly move it to the opposite side with a single step. 

In Cosmic Epoch, if I snap a window to one side, then try to move it to the other side, it first maximizes the window and requires a second keystroke to complete the move to the other side. This adds an extra step to move a window from one side of the monitor to the other. There is already a separate shortcut to maximize the window so this adds unnecessary extra steps. This is important for me because I move windows around with my keyboard often. 

Is there a way to change this behavior or am I missing a shortcut in the documentation?

**Comment by mmstick** [comment]
@avatar4d Sounds like a bug. You can create an issue for it.

**Comment by Drakulix** [comment]
> @avatar4d Sounds like a bug. You can create an issue for it.

Not a bug, but by design.

**Comment by avatar4d** [comment]
@mmstick It didn't seem like a bug to me, but rather that it was by design as indicated. I can see the logic as to why it would be implemented this way for folks not used to keyboard driven interfaces. 

@Drakulix Is there a way to change this behavior, an alternative shortcut, or plans to add either in the future? As much as I want to use Cosmic Epoch as my daily driver, this just doesn't operate as I want, which I guess is like every other window manager's keyboard driven workflow.

**Comment by pietryszak** [comment]
I see that i need to move coursor to change a workspace by shortcut. A something that we discuss months ago. To add a shortcuts to move workpaces by only meta + ctrl + 1/2/3.
Such a missed opurtunity to go to popos from tilling manager like i3. But when I wanna chancge workspace in second monitor i need to move coursor to second monitor and then use shortcut. Pointless

**Comment by RMPR** [comment]
Thank you for making this, I am loving every minute of using Cosmic on Fedora. I am wondering if in addition to <kbd>Super + 0</kbd> to switch to the last workspace we could use <kbd>Super + <workspace number></kbd> to switch to the last workspace. An user flow could look like this: I am on workspace 2, I press <kbd>Super + 4</kbd> to go to workspace 4, if I press <kbd>Super + 4</kbd> again, it should bring me back to Workspace 2. Conversely, if I press <kbd>Super + 2</kbd> while on workspace 2, I should be brought back to workspace 4. I am open to work on the feature if someone can point me in the right direction.

**Comment by pietryszak** [comment]
> Thank you for making this, I am loving every minute of using Cosmic on Fedora. I am wondering if in addition to Super + 0 to switch to the last workspace we could use Super + to switch to the last workspace. An user flow could look like this: I am on workspace 2, I press Super + 4 to go to workspace 4, if I press Super + 4 again, it should bring me back to Workspace 2. Conversely, if I press Super + 2 while on workspace 2, I should be brought back to workspace 4. I am open to work on the feature if someone can point me in the right direction.

`Super + 1`, `Super +2`, `Super + 3` .. change a workspace on monitor number one.
`Super + shift + 1`, `Super + shift +2`, `Super + shift +3`... change workspace on monitor number two.
Like in Sway/i3.

**Comment by anaTropeas** [comment]
changing the keyboard layout is there a keyboard shortcut ?

**Comment by leviport** [comment]
@anaTropeas not yet, but there will be

**Comment by zoof** [comment]
Is it possible to override default keybindings?  When I try to set a keybinding that is already defined, the existing keybinding gets applied and the settings app does not detect my key press.

Otherwise, Cosmic is looking fantastic!

**Comment by nicolasdumitru** [comment]
I think that an optional modal approach to managing windows is a great idea and it could be among the features that take the new Cosmic from pretty good to great. I see that it has also been mentioned by @krmbzds for similar reasons. This is one of the strenghts of the GNOME-based Cosmic. I also think that many (Neo)vim users will likely enjoy the ability to use simpler keybindings (i.e. requiring fewer modifiers and no Super key while in the window adjustment mode) to perform a sequence of multiple changes to the window layout and the ability to have the same key do different things in different modes. Having a single mode where you have to cram all the keybindings will usually result in having either unergonomic (overly complex) or unintuitive keybindings (or both).

**Comment by mipdableep** [comment]
Hi! i hope this is the right thread for this question - is there an option to edit the keyboard shortcuts in a specific text file? (a .toml config file or something similar)

**Comment by tsloughter** [comment]
This seems to be the place to ask, sorry if I should have gone to mattermost. It sounds like there are plans for stack specific keybindings? But one I didn't see mentioned is a binding specific to switching active window in a stack that loops around instead of hopping out of the stack to other windows in the workspace. Would that be possible?

**Comment by Ravengood** [comment]
For me on POP!_OS 22.04 keyboard navigation is broken when i move a window to another workspace or display ... window focus is left behind on the workspace I left (probably focusing last active window) instead of following the window i'm moving to the new workspace (very often i need to move it further or interact with the window, but to do that i need to regain focus) ... is this a bug or a feature?

**Comment by maria-komarova** [comment]
> This seems to be the place to ask, sorry if I should have gone to mattermost. It sounds like there are plans for stack specific keybindings? But one I didn't see mentioned is a binding specific to switching active window in a stack that loops around instead of hopping out of the stack to other windows in the workspace. Would that be possible?

This one hasn't been planned.

@Ravengood are you using COSMIC on 22.04? And it sounds to me like a bug, the focus should be on the window you are manipulating.

**Comment by Ravengood** [comment]
> option to edit the keyboard shortcuts in a specific text file? (a .toml config file or something similar)

> > This seems to be the place to ask, sorry if I should have gone to mattermost. It sounds like there are plans for stack specific keybindings? But one I didn't see mentioned is a binding specific to switching active window in a stack that loops around instead of hopping out of the stack to other windows in the workspace. Would that be possible?
> 
> This one hasn't been planned.
> 
> @Ravengood are you using COSMIC on 22.04? And it sounds to me like a bug, the focus should be on the window you are manipulating.

@maria-komarova ... yes COSMIC on 22.04 ... should i report a new issue regarding this?

**Comment by maria-komarova** [comment]
@Ravengood yes, please, unless similar issue exists already.

**Comment by leviport** [comment]
@Ravengood I believe the behavior your describing matches this issue: https://github.com/pop-os/cosmic-workspaces-epoch/issues/95

**Comment by Ravengood** [comment]
> @Ravengood I believe the behavior your describing matches this issue: [pop-os/cosmic-workspaces-epoch#95](https://github.com/pop-os/cosmic-workspaces-epoch/issues/95)

Absolutely ... thx!

**Comment by mmstick** [comment]
You can use the input sources applet to switch between layouts

**Comment by nirfse** [comment]
As a multilingual user, constant language switching is a natural part of the workflow, much like typing numbers doesn't require manual GUI applet switching - it should be seamless. I appreciate that the current functionality exists, but hope it can be made more streamlined in the future.

**Comment by maria-komarova** [comment]
Super + Space shortcut is planned for input source switching.

**Comment by idoric** [comment]
I am testing COSMIC Epoch for the first time with alpha 5. Here is my feedback as a user of [home row mods](https://precondition.github.io/home-row-mods) and keyboard layers (think [Miryoku](https://github.com/manna-harbour/miryoku) or [Arsenik](https://github.com/OneDeadKey/arsenik)) on a compact keyboard (think [Corne](https://github.com/foostan/crkbd) or [Ferris](https://github.com/pierrechevalier83/ferris)).

When all four modifier keys are on the home row, the Super key is often the one under the little fingers when at rest (at `A` and `;` on a qwerty keyboard). It's not very pleasant to hold down a key with your little finger. There are other HRM configurations that don't have this problem (shift key under one thumb, only three modifier keys on the home row), but they are not the most popular.

My navigation keys are on a navigation layer that I activate with a thumb ([the layer in question](https://github.com/OneDeadKey/arsenik?tab=readme-ov-file#4-navigation-layer)). With most shortcuts this works fine, except when swapping windows. By activating the Super key on my right hand, my left hand remains free for the navigation keys, except when I press X (which is in the same position as on the qwerty keyboard), since the left arrow is assigned to my left ring finger. The problem may also arise for window resizing depending on the key that will be chosen (I read TBD on the board).

I note that the old pop shortcuts (with a more modal approach) didn’t have these problems for this type of keyboard configuration. I therefore join @alapins, @krmbzds and @nicolasdumitru in supporting the wish for an optional modal approach to managing windows.

**Comment by nicolasdumitru** [comment]
@idoric I must say that, after using it for a while and getting used to it, it feels great even as is. I'd still like to have a mode for editing the window layout. Its killer feature would be allowing you to select multiple windows and then operate on them all at once. While it is currently possible to do something similar with "Focus window in/out" keybindings, this requires the windows to be next to each other. Also, as far as window management goes, the main thing I'd argue is missing is a way of moving windows "above" workspace 1 (basically, exactly the same way workspaces are created when you need more and move windows "down", but renumbering existing workspaces).

**Comment by Areopagitics** [comment]
can we add the show/hide desktop (/all windows) command? 

Super + D as default would be really nice

**Comment by iexavl** [comment]
I have a question regarding the order of focus when switching windows with keyboard shortcuts. It seems like right now the windows are focused based completely on their position relative on each other, but this leads to some seemingly (to me at least) weird interactions. Here is an example:

<img width="1913" height="1042" alt="Image" src="https://github.com/user-attachments/assets/f58bfa2c-8be8-4634-96a2-178ada1644fc" />

These two windows are next to each other, but there is a window under them somewhere in the middle. When I have one of the windows on focus and try to focus onto the adjacent one with a shortcut, the window from under them (that is completely obscured) pops out. On the gnome pop-os this behavior is different and it seems to be taking into account how obstructed the windows are when deciding which one should come into focus. This doesn't happen if the window that's presumably under them is minimized.
I am a bit unsure of how I feel about this. On one hand the behavior is more consistent and you just have to minimize the windows underneath and it will work fine. It also doesn't affect tiling mode since all windows are on the screen anyway. However I think having to minimize all windows just because of this can be cumbersome, which is especially true when working with multiple monitors.

**Comment by Areopagitics** [comment]
> can we add the show/hide desktop (/all windows) command?
> 
> Super + D as default would be really nice

Just added a basic implementation (https://github.com/pop-os/cosmic-comp/pull/1969, https://github.com/pop-os/cosmic-settings/pull/1741, https://github.com/pop-os/cosmic-settings-daemon/pull/123)  documented here https://github.com/pop-os/cosmic-comp/issues/1504

It still needs to be made default though.

**Comment by Areopagitics** [comment]
> > can we add the show/hide desktop (/all windows) command?
> > Super + D as default would be really nice
> 
> Just added a basic implementation ([pop-os/cosmic-comp#1969](https://github.com/pop-os/cosmic-comp/pull/1969), [pop-os/cosmic-settings#1741](https://github.com/pop-os/cosmic-settings/pull/1741), [pop-os/cosmic-settings-daemon#123](https://github.com/pop-os/cosmic-settings-daemon/pull/123)) documented here [pop-os/cosmic-comp#1504](https://github.com/pop-os/cosmic-comp/issues/1504)
> 
> It still needs to be made default though.

Now it works almost exactly like Windows, keeping track of which windows are minimized and resetting (see ([pop-os/cosmic-comp#1969](https://github.com/pop-os/cosmic-comp/pull/1969) for the update)

Looking forward to testing and feedback!

**Comment by Thomot512** [comment]
When remapping the shortcuts, most action such as Focus, or move, it is required that you to remap Focus(Left), Focus(Right), ..., or Move(Left), Move(Right), ... individually.

But for example Resizing(Inwards), and Resizing(Outwards) lack such a possibility. 

Could we either have the possibility to remap Resizing(Inwards(Right)), ... Or the possibility to remap Left, Right, Up, Down, In, and Out once and apply to all action such as Move, Focus, Resizing(Inwards), ...?

I hope I'm not hijacking the issue for something it is not supposed to be. If it's the case please tell me and I'll create a new one.

**Comment by DBTDerpbox** [comment]
I'm not sure if this the place to ask (sorry if its not) but, thoughts on a keybind to swap workspaces? Would work similar to Swap Window (Meta+X) but would swap the entire workspace.

**Comment by maria-komarova** [comment]
@Thomot512 this particular shortcut is a bit tricky because it's multi step but we'll see what we can do as we move forward. 

@DBTDerpbox there are just too many shortcuts already so it's increasingly difficult to keep them sensible the more shortcuts are added. You can swap workspaces when using auto-tiling by using a couple of advanced key bindings Super + U and Super + I to select groups of windows and then using Super + X to swap. Super + U will move focus "up" the "window tree" while Super + I will focus windows back down. Not sure if you use automatic tiling so not sure if this would help but there is this functionality.

**Comment by jonathancross** [comment]
In Gnome (22.04) I was very used to thinking of each app as having 1 or more windows.

- **Super Tab** switched between Apps.

- **Super `** switched between separate windows _within a single app_.

In Cosmic (24.04) the model seems to have changed where every window is just a separate entity and there is no clear way to recreate the **Super `** functionality.  Is that correct?  Can we bring back that option?

**Comment by jamochl** [comment]
Is there a way to re-enable the super-enter adjustment mode?

I prefer that way of tiling as it is very vim like, whereas the current way is more emacs like (my hand gets very scrunched up trying to move windows to different monitors)

Plus the adjustment mode makes it easy to adjust window sizes too in the same series of actions as moving them. Which I find very convenient.

**Comment by mmstick** [comment]
@jamochl COSMIC has never supported that feature of the pop-shell gnome-shell extension.

**Comment by jamochl** [comment]
Is it possible to add adjustment mode back as a feature then? The current keybinds can still remain as is, but super+enter can allow people to use adjustment mode if they still wish?
